src/backend/joanie/core/management/commands/create_certificates.py

```diff
@@ -0,0 +1,223 @@
+
+            return str(total)
+
+        if options["courses"]:
+            course_codes = (
+                options["courses"]
+                if isinstance(options["courses"], list)
+                else [options["courses"]]
+            )
+
+        if options["products"]:
+            product_uids = (
+                options["products"]
+                if isinstance(options["products"], list)
+                else [options["products"]]
+            )
+
+        filters = {"type__in": enums.PRODUCT_TYPE_CERTIFICATE_ALLOWED}
+        if course_codes:
+            filters.update({"courses__code__in": course_codes})
+        if product_uids:
+            filters.update({"uid__in": product_uids})
+
+        products = models.Product.objects.filter(**filters)
+
+        for product in products:
+            result = self._generate_certificate_for_product(product, course_codes)
+            total += result
+
+        return str(total)
+
+    def _generate_certificate_for_product(self, product, course_codes):
+        """
+        Retrieve orders related to the product and course code then check if
+        it is eligible for certification.
+        """
+        filters = {"is_canceled": False, "certificate__isnull": True}
+        if course_codes:
+            filters.update({"course__code__in": course_codes})
+
+        orders = [
+            order
+            for order in product.orders.filter(**filters).select_related(
+                "course__organization"
+            )
+            if order.state == enums.ORDER_STATE_VALIDATED
+        ]
+
+        certificate_counter = 0
+        for order in orders:
+            result = self._generate_certificate_for_order(order)
+            certificate_counter += result
+
+        if certificate_counter > 0:
+            logger.info(
+                '%s certificate(s) for "%s" have been generated.',
+                certificate_counter,
+                product.title,
+            )
+
+        return certificate_counter
+
+    @staticmethod
+    def _generate_certificate_for_order(order):
+        """
+        Check if order is eligible for certification then generate certificate is it is.
+
+        Eligibility means that order contains
+        one passed enrollment per graded courses.
+
+        Return:
+            0 : if the order is not eligible to certification
+            1: if a certificate has been generated for the current order
+        """
+        graded_courses = order.target_courses.filter(
+            order_relations__is_graded=True
+        ).order_by("order_relations__position")
+        graded_courses_count = len(graded_courses)
+
+        if graded_courses_count == 0:
+            return 0
+
+        course_runs = models.CourseRun.objects.filter(
+            course__in=graded_courses,
+            is_gradable=True,
+            start__lte=timezone.now(),
+        )
+
+        enrollments = order.enrollments.filter(
+            course_run__in=course_runs, is_active=True
```

Comment:
there is a small trick here: you say "generate for order" but you may select enrollments here that are related to other orders staging the same course run... :thinking: Add a filter clause and a test if I'm right?